content/news/2022-04-12-bevy-0.7/index.md

```diff
@@ -0,0 +1,383 @@
+    }
+}
+```
+
+But ParamSets aren't just limited to Queries! Consider this example, where the `EventWriter<Jump>` parameter (which internally accesses the `Events<Jump>` resource) conflicts with the raw access to that resource. Previously, expressing this wouldn't be possible. But with ParamSets, it is!
+
+```rust
+fn system(mut set: ParamSet<(EventWriter<Jump>, ResMut<Events<Jump>>)>) {
+    for jump_event in set.p1().drain() {
+    }
+}
+```
+
+We still recommend avoiding ParamSets where possible for clarity's sake. But every so often they are a necessary and useful tool!
+
+## Deref / DerefMut Derives
+
+<div class="release-feature-authors">authors: @MrGVSV</div>
+
+Rust encourages the use of the [newtype pattern](https://doc.rust-lang.org/rust-by-example/generics/new_types.html) when expanding a type with new functionality or meaning. This is also a useful tool in Bevy:
+
+```rust
+#[derive(Component)]
+struct Items(Vec<Item>);
+
+fn give_sword(mut query: Query<&mut Items>) { 
+    for mut items in query.iter_mut() {
+        items.0.push(Item::new("Flaming Poisoning Raging Sword of Doom"));
+    }
+}
+```
+
+This works just fine, but that `0` at the end of `items.0` sticks out like a sore thumb. Many of us in the Bevy Org think `.0` has no place in public apis. But the newtype pattern is still useful! Ideally, Rust would provide a way to express that `Items` is a new type, while transparently provided access to the `Vec<Item>` stored within. There are designs being discussed by the Rust team, but we don't want to wait for nice things!
+
+Fortunately, the Deref / DerefMut traits in std provide the behavior we want. Users can already manually implement these traits, but for such a common pattern, we decided that providing our own trait derives was worth it. In **Bevy 0.7**, you can now derive Deref and DerefMut, enabling much nicer public apis:
+
+```rust
+#[derive(Component, Deref, DerefMut)]
+struct Items(Vec<Item>);
+
+fn give_sword(mut query: Query<&mut Items>) { 
+    for mut items in query.iter_mut() {
+        // No more .0!
+        items.push(Item::new("Flaming Poisoning Raging Sword of Doom"));
+    }
+}
+```
+
+Astute `std` doc readers might notice that the Rust team [recommends only using Deref/DerefMut for smart pointers, to avoid confusion](https://doc.rust-lang.org/std/ops/trait.Deref.html). Components like `Items` _are not_ smart pointers. We personally don't care. This pattern works, it is already widely used in the Rust ecosystem, and Good UX Comes First.
+
+## WorldQuery Derives
+
+<div class="release-feature-authors">authors: @mvlabat</div>
+
+Sometimes when building Bevy Apps you might find yourself repeating the same sets of components over and over in your queries:
+
+```rust
+fn move_players(mut players: Query<(&mut Name, &mut Transform, &mut Item)>) {
+    for (name, transform, item) in players.iter_mut() {
+    }
+}
+
+fn despawn_players(mut players: Query<(Entity, &mut Name, &mut Transform, &mut Item)>) {
```

Comment:
I'm not a huge fan of this example, since for despawning, it seems unlikely that you'd need full access to all of these components, and certainly not for the same reason as in `move_players`

Unfortunately, I'm not sure of any better examples. Maybe this is almost always unidiomatic, but we provide it for completeness?

Reply:
I have a real-world example where this is useful. It's not "unidiomatic".

Say, you have a "player controller" that is represented as multiple components, like

```rust
/// Player controller configuration/properties
struct PlayerControlParams {
    gravity_up: f32,
    gravity_down: f32,
    acceleration_start: f32,
    acceleration_end: f32,
    // ...
}
/// Current player movement state
enum PlayerMoveState {
    Standing,
    Walking,
    Jumping,
    Falling,
}
/// Desired movement, as per joystick or other input
struct PlayerWantsMovement(Vec2);
```

These are separate components, because some systems (like the "joystick input" or whatever) might only need specific ones. But systems that are part of the player controller would typically need all of them.

```rust
#[derive(WorldQuery)]
struct PlayerControlQuery<'w> {
    transform: &'w mut Transform,
    state: &'w mut PlayerMoveState,
    input: &'w PlayerWantsMovement,
    params: &'w PlayerControlParams,
}

fn player_gravity(q: Query<PlayerControlQuery>) { /* ... */ }
fn player_walk(q: Query<PlayerControlQuery>) { /* ... */ }
fn input_gamepad(q: Query<&mut PlayerWantsMovement>, /* ... */) { /* ... */ }
fn input_keyboard(q: Query<&mut PlayerWantsMovement>, /* ... */) { /* ... */ }
```

Now, this is probably *too much* to include as an example in the release blog post. I just wanted to show a real use case for this. Feel free to use it as inspiration and come up with something more concise.

Reply:
Cool yeah I can definitely make this a bit less arbitrary.

Reply:
I drew inspiration from @inodentry's example and made this a bit less arbitrary.